v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/transforms/SourceWriterFn.java

```diff
@@ -114,37 +115,30 @@ public void setSpannerDao(SpannerDao spannerDao) {
   }
 
   // for unit testing purposes
-  public void setMySqlDaoMap(Map<String, MySqlDao> mySqlDaoMap) {
-    this.mySqlDaoMap = mySqlDaoMap;
+  public void setObjectMapper(ObjectMapper mapper) {
+    this.mapper = mapper;
   }
 
   // for unit testing purposes
-  public void setObjectMapper(ObjectMapper mapper) {
-    this.mapper = mapper;
+  public void setSourceProcessor(SourceProcessor sourceProcessor) {
+    this.sourceProcessor = sourceProcessor;
   }
 
   /** Setup function connects to Cloud Spanner. */
   @Setup
-  public void setup() {
+  public void setup() throws UnsupportedSourceException {
     mapper = new ObjectMapper();
     mapper.enable(DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS);
-    ConnectionHelper.init(shards, null, maxThreadPerDataflowWorker);
-    mySqlDaoMap = new HashMap<>();
-    for (Shard shard : shards) {
-      String sourceConnectionUrl =
-          "jdbc:mysql://" + shard.getHost() + ":" + shard.getPort() + "/" + shard.getDbName();
-      mySqlDaoMap.put(
-          shard.getLogicalShardId(),
-          new MySqlDao(sourceConnectionUrl, shard.getUserName(), shard.getPassword()));
-    }
+    sourceProcessor =
+        SourceProcessorFactory.createSourceProcessor(source, shards, maxThreadPerDataflowWorker);
     spannerDao = new SpannerDao(spannerConfig);
   }
 
   /** Teardown function disconnects from the Cloud Spanner. */
   @Teardown
   public void teardown() throws Exception {
     spannerDao.close();
-    mySqlDaoMap.clear();
+    sourceProcessor.getSourceDaoMap().clear();
```

Comment:
This is a leaky abstraction. Have cleanup/close or some such method in the SourceProcessor that cleans up it's internal state. The caller need not be aware that there is a map in the source processor that needs to be cleaned.